File: javaSpringShip/src/main/java/pl/battleships/javaspringship/client/RestTemplateResponseErrorHandler.java
package pl.battleships.javaspringship.client;

import org.springframework.http.client.ClientHttpResponse;
import org.springframework.web.client.ResponseErrorHandler;
import pl.battleships.core.exception.GameOverException;
import pl.battleships.core.exception.InvalidMoveException;
import pl.battleships.core.exception.NoGameFoundException;

import java.io.IOException;

public class RestTemplateResponseErrorHandler implements ResponseErrorHandler {
    @Override
    public boolean hasError(ClientHttpResponse response) throws IOException {
        return !response.getStatusCode().is2xxSuccessful();
    }

    @Override
    public void handleError(ClientHttpResponse response) throws IOException {
        switch (response.getStatusCode()) {
            case NOT_FOUND:
                throw new NoGameFoundException();
            case GONE:
                throw new GameOverException();
            case FORBIDDEN:
                throw new InvalidMoveException();
        }
    }
}

File: kotlinSpringShip/src/main/kotlin/pl/battleships/kotlinspringship/client/RestTemplateResponseErrorHandler.kt
package pl.battleships.kotlinspringship.client

import org.springframework.http.HttpStatus
import org.springframework.http.client.ClientHttpResponse
import org.springframework.web.client.ResponseErrorHandler
import pl.battleships.core.exception.GameOverException
import pl.battleships.core.exception.InvalidMoveException
import pl.battleships.core.exception.NoGameFoundException

class RestTemplateResponseErrorHandler : ResponseErrorHandler {
    override fun hasError(response: ClientHttpResponse): Boolean {
        return !response.statusCode.is2xxSuccessful
    }

    override fun handleError(response: ClientHttpResponse) {
        when (response.statusCode) {
            HttpStatus.NOT_FOUND -> throw NoGameFoundException()
            HttpStatus.GONE -> throw GameOverException()
            HttpStatus.FORBIDDEN -> throw InvalidMoveException()
        }
    }
}